Code/Framework/AzCore/AzCore/DOM/DomValue.h

```diff
@@ -289,8 +289,8 @@ namespace AZ::Dom
         void RemoveMember(KeyType name);
         void RemoveMember(AZStd::string_view name);
         Object::Iterator RemoveMember(Object::Iterator pos);
-        Object::Iterator EraseMember(Object::ConstIterator pos);
-        Object::Iterator EraseMember(Object::ConstIterator first, Object::ConstIterator last);
+        Object::Iterator EraseMember(Object::Iterator pos);
+        Object::Iterator EraseMember(Object::Iterator first, Object::Iterator last);
```

Comment:
A const iterator should work just as well fro removing elements from an Array or Object.

The iterator being const doesn't affect the constness of the `DomValue` itself it is reasing an object from itself.

C++ standard Container class allow erasure through a [const iterator,](https://en.cppreference.com/w/cpp/string/basic_string/erase) because it is not the iterator that is being modified, it is the container.

Reply:
We explicitly disallow it here because of copy-on-write. If you get a const iterator to a shared `Value` container, then attempt to mutate it, the container storage pointer changes out from underneath you and your iterator is no longer valid. To prevent this, I require a mutating iterator, which will force a detach at iterator lookup time.

Reply:
That doesn't make any sense, because erasing an element at the iterator position must invalidate the iterator.

So their is no need for a mutating iterator.

Reply:
The order of operations right now is:
-Request for erase comes in
-Internally, we check the copy on write state of the container. If there's more than one reference, detach and create a new container
-Perform the erase operation. If the container has been copied, we're no longer a valid iterator inside the container.

This API change prevents that from occurring by requiring a non-const iterator, which will have already forced the detach.

We could also fix up the iterators post-hoc by migrating indices, but I didn't prioritize it as it's just as easy to grab a non-const iterator. Does that help clarify the situation?

Reply:
The iterator shouldn't be valid even if the container wasn't copied.

We are not mutating the iterator.

Reply:
We're not mutating the iterator, but importantly the iterator becomes invalid **before** the erase operation occurs, thus causing it to crash, so unless mutating methods that take an iterator migrate the const iterator to a valid iterator in the new container, this will always occur.

I'm not expecting this to be a usability issue (if you're calling Erase you can always obtain a non-const iterator) but if you disagree I can spend a bit of time making const iterators work.